dv/cosim/spike_cosim.cc

```diff
@@ -0,0 +1,516 @@
+    // true, otherwise assume a dside access and check against DUT dside
+    // accesses.
+    dut_error = check_mem_access(false, addr, len, bytes);
+  }
+
+  return !(bus_error | dut_error);
+}
+
+bool SpikeCosim::mmio_store(reg_t addr, size_t len, const uint8_t *bytes) {
+  bool bus_error = !bus.store(addr, len, bytes);
+  bool dut_error = check_mem_access(true, addr, len, bytes);
+
+  return !(bus_error | dut_error);
+}
+
+void SpikeCosim::proc_reset(unsigned id) {}
+
+const char *SpikeCosim::get_symbol(uint64_t addr) { return nullptr; }
+
+void SpikeCosim::add_memory(uint32_t base_addr, size_t size) {
+  auto new_mem = std::make_unique<mem_t>(size);
+  bus.add_device(base_addr, new_mem.get());
+  mems.emplace_back(std::move(new_mem));
+}
+
+bool SpikeCosim::backdoor_write_mem(uint32_t addr, size_t len,
+                                    const uint8_t *data_in) {
+  return bus.store(addr, len, data_in);
+}
+
+bool SpikeCosim::backdoor_read_mem(uint32_t addr, size_t len,
+                                   uint8_t *data_out) {
+  return bus.load(addr, len, data_out);
+}
+
+bool SpikeCosim::step(uint32_t write_reg, uint32_t write_reg_data, uint32_t pc,
+                      bool sync_trap) {
+  assert(write_reg < 32);
+
+  if (pc_is_mret(pc)) {
+    nmi_mode = false;
+  }
+
+  uint32_t initial_pc = (processor->get_state()->pc & 0xffffffff);
+  bool initial_pc_match = initial_pc == pc;
+
+  // Execute the next instruction
+  processor->step(1);
+
+  if (processor->get_state()->last_inst_pc == PC_INVALID) {
+    if (processor->get_state()->mcause->read() & 0x80000000) {
+      // Interrupt occurred, step again to execute first instruction of
+      // interrupt
+      processor->step(1);
+      // TODO: Deal with exception on first instruction of interrupt
+      assert(processor->get_state()->last_inst_pc != PC_INVALID);
+    } else {
+      // Otherwise a synchronous trap has occurred, check the DUT reported a
+      // synchronous trap at the same point
+      if (!sync_trap) {
+        std::stringstream err_str;
+        err_str << "Synchronous trap was expected at ISS PC: " << std::hex
+                << processor->get_state()->pc
+                << " but DUT didn't report one at PC " << pc;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      if (!initial_pc_match) {
+        std::stringstream err_str;
+        err_str << "PC mismatch at synchronous trap, DUT: " << std::hex << pc
+                << " expected: " << std::hex << initial_pc;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      if (write_reg != 0) {
+        std::stringstream err_str;
+        err_str << "Synchronous trap occurred at PC: " << std::hex << pc
+                << "but DUT wrote to register: x" << std::dec << write_reg;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      return errors.size() == 0;
```

Comment:
Is there any way for `errors.size()` to be nonzero if we didn't return `false` above? If not, maybe an assertion and then an explicit `return true` would be better?

Reply:
Yes, `check_mem_accesses` adds to `errors`. In particular for stores if there's a bad address/be/data that won't be checked by anything in `step` itself.

In general the idea was anything can add to `errors` and `step` will report failure if so. For now it's just `check_mem_accesses` that does this.

Reply:
Ah! Makes sense. Maybe it would be worth adding a comment that explains how this works?